### PR TITLE
odo dev on podman: Add support for devfile volume

### LINKED ISSUE #6330: Support devfile volumes during dev on podman
/kind feature

## Which functionality do you think we should add?

Volumes and volume mounts mounts defined in a Devfile should be defined in the Pod deployed on Podman during `odo dev --run-on podman`

## Why is this needed?

Some Devfiles in the Devfile registry are defining volumes

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #6330

**PR acceptance criteria:**

- [x] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: pkg/dev/podmandev/pod.go [hunk 1]
```diff
@@ -58,6 +58,26 @@ func createPodFromComponent(
...
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, devfileVolume := range devfileVolumes {
+		volumes = append(volumes, corev1.Volume{
+			Name: devfileVolume.Name,
+			VolumeSource: corev1.VolumeSource{
+				PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+					ClaimName: getVolumeName(componentName, appName, devfileVolume.Name),
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/dev/podmandev/pod.go:71]: You are right, I think misunderstood the naming strategy on cluster. I have made chages, to change name for devfile volumes and also for the "shared" volume. WDYT?

File: pkg/dev/podmandev/pod.go [hunk 2]
```diff
@@ -77,7 +97,7 @@ func createPodFromComponent(
 }
 
 func getVolumeName(componentName string, appName string, volume string) string {
```
Comment (valaparthvi) [inline pkg/dev/podmandev/pod.go:99]: nit:
```suggestion
func getVolumeName(volume string, componentName string, appName string) string {
```

File: pkg/dev/podmandev/pod.go [hunk 3]
```diff
@@ -44,20 +44,40 @@ func createPodFromComponent(
 			Name: storage.OdoSourceVolume,
 			VolumeSource: corev1.VolumeSource{
 				PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
-					ClaimName: getVolumeName(componentName, appName, "source"),
+					ClaimName: getVolumeName(componentName, appName, "odo-projects"),
```
Comment (valaparthvi) [inline pkg/dev/podmandev/pod.go:47]:
```suggestion
					ClaimName: getVolumeName(componentName, appName, storage.OdoSourceVolume"),
```

File: pkg/dev/podmandev/pod.go [hunk 4]
```diff
@@ -44,20 +44,40 @@ func createPodFromComponent(
...
 				},
 			},
 		},
 		{
 			Name: storage.SharedDataVolumeName,
 			VolumeSource: corev1.VolumeSource{
 				PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
-					ClaimName: getVolumeName(componentName, appName, "shared"),
+					ClaimName: getVolumeName(componentName, appName, "odo-shared"),
```
Comment (valaparthvi) [inline pkg/dev/podmandev/pod.go:55]:
```suggestion
					ClaimName: getVolumeName(componentName, appName, storage.SharedDataVolumeName),
```

**Review by valaparthvi — CHANGES_REQUESTED** [review]

File: pkg/dev/podmandev/pod.go (hunk 4 above)
Comment (valaparthvi) [reply pkg/dev/podmandev/pod.go:55]: I am not sure when/where/how we use shared volume though. It is supposed to be the volume shared between InitContainer and the applicationContainer, but I don't know how init containers can be defined inside Devfile and used with `odo dev`.

File: pkg/dev/podmandev/pod.go [hunk 5]
```diff
@@ -44,20 +44,40 @@ func createPodFromComponent(
...
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, devfileVolume := range devfileVolumes {
+		volumes = append(volumes, corev1.Volume{
+			Name: devfileVolume.Name,
+			VolumeSource: corev1.VolumeSource{
+				PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+					ClaimName: getVolumeName(componentName, appName, devfileVolume.Name),
```
Comment (feloy) [inline pkg/dev/podmandev/pod.go:71]:
```suggestion
					ClaimName: getVolumeName(devfileVolume.Name, componentName, appName),
```

**Review by valaparthvi — APPROVED** [review]

**Review by rm3l — COMMENTED** [review]
Wondering if we shouldn't generate `EmptyDir` volumes when the `Ephemeral` preference is `true`, to be consistent with the cluster mode?
I just saw that Podman could support `EmptyDir` volumes by creating anonymous volumes (https://github.com/containers/podman/pull/15473).
(But if you prefer, we can handle this in a separate issue/PR as well).

**Comment by feloy** [comment]
> Wondering if we shouldn't generate `EmptyDir` volumes when the `Ephemeral` preference is `true`, to be consistent with the cluster mode? I just saw that Podman could support `EmptyDir` volumes by creating anonymous volumes ([containers/podman#15473](https://github.com/containers/podman/pull/15473)). (But if you prefer, we can handle this in a separate issue/PR as well).

Effectively, podman can handle emptyDir and PVC, and they are instantiated practically the same way: both are podman volumes, and emptyDir ones will be removed when pod is removed, as opposed to the PVC will be removed by odo at the same time the pod is removed.

I didn't want to add complexity for this small difference. But we can handle this in a separate PR is necessary

**Review by rm3l — APPROVED** [review]
> > Wondering if we shouldn't generate `EmptyDir` volumes when the `Ephemeral` preference is `true`, to be consistent with the cluster mode? I just saw that Podman could support `EmptyDir` volumes by creating anonymous volumes ([containers/podman#15473](https://github.com/containers/podman/pull/15473)). (But if you prefer, we can handle this in a separate issue/PR as well).
> 
> Effectively, podman can handle emptyDir and PVC, and they are instantiated practically the same way: both are podman volumes, and emptyDir ones will be removed when pod is removed, as opposed to the PVC will be removed by odo at the same time the pod is removed.
> 
> I didn't want to add complexity for this small difference. But we can handle this in a separate PR is necessary

Okay, no problem. I added a reminder note about this in the parent Epic.

/approve

**Comment by feloy** [comment]
/override ci/prow/v4.11-integration-e2e
Pass on IBM Cloud

**Pull request merged**